packages/shell-api/src/shell-internal-state.ts

```diff
@@ -0,0 +1,238 @@
+import {
+  AggregationCursor,
+  CursorIterationResult,
+  Cursor,
+  Help,
+  Database,
+  Mongo,
+  ReplicaSet,
+  Shard,
+  signatures,
+  ShellBson,
+  toIterator,
+} from './index';
+import { EventEmitter } from 'events';
+import { DatabaseOptions, Document, ServiceProvider, ReplPlatform } from '@mongosh/service-provider-core';
+import { MongoshInvalidInputError, MongoshUnimplementedError } from '@mongosh/errors';
+import AsyncWriter from '@mongosh/async-rewriter';
+
+/**
+ * Anything to do with the internal shell state is stored here.
+ */
+export default class ShellInternalState {
+  public currentCursor: Cursor | AggregationCursor;
+  public currentDb: Database;
+  public messageBus: EventEmitter;
+  public asyncWriter: AsyncWriter;
+  public initialServiceProvider: ServiceProvider; // the initial service provider
+  public uri: string;
+  public connectionInfo: any;
+  constructor(initialServiceProvider: ServiceProvider, messageBus: any = new EventEmitter()) {
+    this.initialServiceProvider = initialServiceProvider;
+    this.messageBus = messageBus;
+    this.asyncWriter = new AsyncWriter(signatures);
+    const mongo = new Mongo(this);
+    this.currentCursor = null;
+    this.currentDb = mongo.getDB('test'); // TODO: set to CLI arg
+  }
+
+  async fetchConnectionInfo(): Promise<void> {
+    this.connectionInfo = await this.currentDb.mongo.serviceProvider.getConnectionInfo();
+    this.messageBus.emit('mongosh:connect', this.connectionInfo.extraInfo);
+  }
+
+  // TODO: do we need to close all Mongos ever made?
+  close(p): void {
+    try {
+      this.currentDb.mongo.serviceProvider.close(p);
+      // eslint-disable-next-line no-empty
+    } catch (e) {
+    }
+  }
+
+  use(db): any {
+    return this.currentDb.mongo.use(db);
+  }
+  show(arg): any {
+    return this.currentDb.mongo.show(arg);
+  }
+  async it(): Promise<any> {
+    if (!this.currentCursor) {
+      // TODO: warn here
+      return new CursorIterationResult();
+    }
+    return await this.currentCursor._it();
+  }
+  public help(): Help {
```

Comment:
I think with a bit of creativity we could generate this help method too.

Reply:
How would you write a decorator that could generate an attribute whose name is 'show databases' or '.exit'?

Reply:
I've moved these methods into a ShellApi class in this commit: https://github.com/mongodb-js/mongosh/pull/195/commits/b6fdfd1220769050694d7a7b4f44c3bd7771526f

Reply:
The first thing that comes to mind is something like this: 

```
@shellApiDefaultClass({ additionalHelpKeys: ['show databases', '.exit'] })
class SomeClass {}
```

we can do that later, in a dedicated story.

Reply:
Ticketed here: https://jira.mongodb.org/browse/MONGOSH-215